internal/app/cache/cache.go

```diff
@@ -1,19 +1,134 @@
+	return &cache{
+		cache: lru.Cache{
+			// Max number of cache entries before an item is evicted. Zero means no limit.
+			MaxEntries: maxEntries,
+			// OnEvict is called for each eviction and closes the stream if a key is removed
+			// (e.g. expiry due to TTL, too many entries).
+			OnEvicted: onEvicted,
+		},
+		// Duration before which an item is evicted for expiring. Zero means no expiration time.
+		ttl: ttl,
+	}
+}
+
+func (c *cache) Fetch(key string) (*v2.DiscoveryResponse, error) {
+	c.cacheMu.Lock()
+	defer c.cacheMu.Unlock()
+	value, found := c.cache.Get(key)
+	if !found {
+		return nil, fmt.Errorf("no value found for key: %s", key)
+	}
+	resource, ok := value.(resource)
+	if !ok {
+		return nil, fmt.Errorf("unable to cast cache value to type resource for key: %s", key)
+	}
+	// Lazy eviction based on TTL occurs here. Fetch does not increase the lifespan of the key.
+	if resource.isExpired() {
+		c.cache.Remove(key)
+		return nil, nil
+	}
+	return resource.resp, nil
+}
+
+func (c *cache) SetResponse(key string, resp v2.DiscoveryResponse) ([]*v2.DiscoveryRequest, error) {
+	c.cacheMu.Lock()
+	defer c.cacheMu.Unlock()
+	value, found := c.cache.Get(key)
+	if !found {
+		resource := resource{
+			resp:           &resp,
+			expirationTime: c.getExpirationTime(),
+		}
+		c.cache.Add(key, resource)
+		return nil, nil
+	}
+	resource, ok := value.(resource)
+	if !ok {
+		return nil, fmt.Errorf("unable to cast cache value to type resource for key: %s", key)
+	}
+	resource.resp = &resp
+	resource.expirationTime = c.getExpirationTime()
+	c.cache.Add(key, resource)
+	// TODO: Add logic that allows for notifying of watches.
+	return resource.requests, nil
+}
+
+func (c *cache) AddRequest(key string, req v2.DiscoveryRequest) (bool, error) {
+	c.cacheMu.Lock()
+	defer c.cacheMu.Unlock()
+	value, found := c.cache.Get(key)
+	if !found {
+		// TODO: Add logic to guarantee that a stream has been opened.
+		resource := resource{
+			requests:       []*v2.DiscoveryRequest{&req},
+			streamOpen:     true,
+			expirationTime: c.getExpirationTime(),
+		}
+		c.cache.Add(key, resource)
+		return true, nil
+	}
+	resource, ok := value.(resource)
+	if !ok {
+		return false, fmt.Errorf("unable to cast cache value to type resource for key: %s", key)
+	}
+	resource.requests = append(resource.requests, &req)
+	resource.expirationTime = c.getExpirationTime()
+	// TODO: Add logic to guarantee that a stream has been opened.
+	resource.streamOpen = true
+	c.cache.Add(key, resource)
+	return true, nil
+}
+
+func (r *resource) isExpired() bool {
+	if r.expirationTime.IsZero() {
+		return false
+	}
+	return r.expirationTime.Before(time.Now())
+}
 
-	// AddWatch adds the watch to the cache and returns whether a stream is open.
-	AddWatch(key string, req envoy_api_v2.DiscoveryRequest) (bool, error)
+func (c *cache) getExpirationTime() time.Time {
+	if c.ttl > 0 {
+		return time.Now().Add(c.ttl)
+	}
+	return time.Time{}
```

Comment:
what is the default value? time.Now?

Reply:
The default value is time.Time's 0 value. From https://golang.org/pkg/time/#Time:

> The zero value of type Time is January 1, year 1, 00:00:00.000000000 UTC. As this time is unlikely to come up in practice, the IsZero method gives a simple way of detecting a time that has not been initialized explicitly.

`IsZero` is used to determine if expiration is disabled: https://github.com/envoyproxy/xds-relay/blob/a3ff4be3d6288f2982396770d38335059ad835c8/internal/app/cache/cache.go#L123